### PR TITLE
Feature to download model instance

### PR DESCRIPTION
**Pull request by appukuttan-shailesh**
Allows downloading of model instance source files. Following scenarios have been handled:
(source, sample **DEV** model instance UUIDs indicated)

**1) Collab Storage**

	1.1) Path to file
		e.g. https://collab.humanbrainproject.eu/#/collab/5165/nav/39966?state=uuid%3D95672f5e-ce23-4da1-9c4b-74bbee315c9a
		model instance UUID = 717560cf-f9d8-4c42-a4e2-a67bef2aa5cf

	1.2) Path to folder
		e.g. https://collab.humanbrainproject.eu/#/collab/5165/nav/39966?state=uuid%3D522cc44f-6af8-4de2-9084-01ff498b2d53
		model instance UUID = e133fb81-1197-4e1d-89e7-69065c766272

**2) Private CSCS container**

	2.1) Path to file
		e.g. swift://cscs.ch/SGA1_basal_ganglia_private/data/Electrophysiology/str-chin-jofn20180320/Cell1_2/md_20160504_cell_1_2_ChIN.mat
		model instance UUID = 2eb29985-6c2c-49e6-8d49-8d383b88585e

	2.2) Path to folder
		e.g. swift://cscs.ch/SGA1_basal_ganglia_private/data/Electrophysiology/str-chin-jofn20180320/Cell1_2
		model instance UUID = b71ea090-41dc-40bb-97fb-29034512183c

**3) Pubic CSCS container**

	3.1) Path to file
		e.g. https://object.cscs.ch/v1/AUTH_c0a333ecf7c045809321ce9d9ecdfdea/hippocampus_optimization/optimizations/CA1_int_cAC_990611HP2_20180122160209/CA1_int_cAC_990611HP2_20180122160209.zip
		model instance UUID = 6db05cb6-8e61-4bb4-be79-ce9a6344bca2

	3.2) Path to folder
		e.g. https://object.cscs.ch/v1/AUTH_c0a333ecf7c045809321ce9d9ecdfdea/hippocampus_optimization/optimizations/CA1_int_cAC_990611HP2_20180122160209
		model instance UUID = 8cf18126-3085-4d7a-8805-937d7a7dab27

**4) External url to file**

	4.1) IPython notebook
		e.g. http://nbviewer.jupyter.org/url/docs.thevirtualbrain.org/tutorials/tutorial_s6_ModelingEpilepsy.ipynb
		model instance UUID = ee35d0b7-b7c4-4e44-932d-a43fd55cdc2a

	4.2) Zip File
		e.g. http://www.neuroconstruct.org/models/downloads/CA1PyramidalCell.ncx.zip
		model instance UUID = 0fa6811e-ce1f-4927-9d09-081295d9c1b5

	4.3) ModelDB
		e.g. https://senselab.med.yale.edu/modeldb/eavBinDown.cshtml?o=240957&a=23&mime=application/zip
		model instance UUID = 47734fd5-1304-4d09-b884-2097b09d6e1a

### CODE REVIEW

**Comment by appukuttan-shailesh** [comment]
Test code for above:
```
from hbp_validation_framework import ModelCatalog
mc = ModelCatalog(username="ENTER_USERNAME", environment="dev")

instance_uuids = ["717560cf-f9d8-4c42-a4e2-a67bef2aa5cf", "e133fb81-1197-4e1d-89e7-69065c766272", "2eb29985-6c2c-49e6-8d49-8d383b88585e", "b71ea090-41dc-40bb-97fb-29034512183c", "6db05cb6-8e61-4bb4-be79-ce9a6344bca2", "8cf18126-3085-4d7a-8805-937d7a7dab27", "ee35d0b7-b7c4-4e44-932d-a43fd55cdc2a", "0fa6811e-ce1f-4927-9d09-081295d9c1b5", "47734fd5-1304-4d09-b884-2097b09d6e1a"]

for i in range(len(instance_uuids)):
	print "-------------------------"
	print mc.download_model_instance(instance_id=instance_uuids[i], target_loc="./DOWNLOAD/LOC")
```

File: hbp_validation_framework/__init__.py [hunk 1]
```diff
@@ -338,10 +344,12 @@ def _download_resource(self, uuid):
             raise Exception("The provided 'uuid' is invalid!")
         else:
             data = data.json()
+            dirpath = target_loc
             if data["entity_type"] == "folder":
-                if not os.path.exists(data["name"]):
-                    os.makedirs(data["name"])
-                os.chdir(data["name"])
+                dirpath = os.path.join(target_loc, data["name"])
+                if not os.path.exists(dirpath):
```
Comment (apdavison) [inline hbp_validation_framework/__init__.py]: maybe use the `Path` module so as to support the `exists_ok` option. (what if `target_loc` is a deeply nested path where some of the intermediate directories don't exist?)

File: hbp_validation_framework/__init__.py [hunk 2]
```diff
@@ -338,10 +344,12 @@ def _download_resource(self, uuid):
...
+                    os.makedirs(dirpath)
+                os.chdir(dirpath)
```
Comment (apdavison) [inline hbp_validation_framework/__init__.py]: I don't understand why you are changing directory here.

File: hbp_validation_framework/__init__.py [hunk 3]
```diff
@@ -1910,6 +1921,152 @@ def get_model_instance(self, instance_path="", instance_id="", model_id="", alia
         else:
             raise Exception("Error in retrieving model instance. Possibly invalid input data.")
 
+    def download_model_instance(self, instance_path="", instance_id="", model_id="", alias="", version="", target_loc="."):
```
Comment (apdavison) [inline hbp_validation_framework/__init__.py]: Same comment as above about `target_loc`

File: hbp_validation_framework/__init__.py [hunk 4]
```diff
@@ -323,13 +328,14 @@ def _translate_URL_to_UUID(self, path):
         else:
             raise Exception("Error: " + data.content)
 
-    def _download_resource(self, uuid):
+    def _download_resource(self, uuid, target_loc="."):
```
[the remark this answers is not in the document]
Comment (apdavison) [reply hbp_validation_framework/__init__.py]: Also, this function seems to be redundant with the `datastores.CollabDataStore.download_data` method.

File: hbp_validation_framework/__init__.py [hunk 5]
```diff
@@ -1910,6 +1921,152 @@ def get_model_instance(self, instance_path="", instance_id="", model_id="", alia
...
+        """
+
+        model_source = self.get_model_instance(instance_path=instance_path, instance_id=instance_id, model_id=model_id, alias=alias, version=version)["source"]
+        if model_source[-1]=="/":
+            model_source = model_source[:-1]    # remove trailing '/'
+        if not os.path.exists(target_loc):
+            os.makedirs(target_loc) # create directory if it doesn't exist
+        fileList = []
+
+        if model_source.startswith("https://collab.humanbrainproject.eu/#/collab/"):
```
Comment (apdavison) [inline hbp_validation_framework/__init__.py]: In the interests of ease-of-testing, maintainability and extensibility, I strongly suggest moving this functionality into the datastores module, i.e.

- extend `CollabDataStore`, if necessary
- add a `SwiftDataStore` class.

File: hbp_validation_framework/__init__.py (hunk 1 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/__init__.py]: I think `os.makedirs` takes care of any missing intermediate directories:
https://docs.python.org/2/library/os.html#os.makedirs

But yes the `Path` module seems more appropriate for this purpose. Will shift to this.

File: hbp_validation_framework/__init__.py (hunk 5 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/__init__.py]: Will look into this and see how best it to accommodate it there.

File: hbp_validation_framework/__init__.py (hunk 1 above)
Comment (apdavison) [reply hbp_validation_framework/__init__.py]: Sorry, I mixed things up. The problem with `os.makedirs` is when the final directory already exists, but this is handled by your `os.path.exists` call

Comment (appukuttan-shailesh) [reply hbp_validation_framework/__init__.py]: Decided to implement all directory creation using `Path` module. Changes made accordingly.

File: hbp_validation_framework/__init__.py (hunk 2 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/__init__.py]: Yes, this was avoidable. Have updated the target directory specification such that this is no longer required.

File: hbp_validation_framework/__init__.py (hunk 5 above)
Comment (appukuttan-shailesh) [reply hbp_validation_framework/__init__.py]: Done as per above suggestion. Also moved a few methods from __init__.py to CollabDataStore (ideally I should never have placed them in the former). So a good amount of code clean-up carried out.

**Comment by appukuttan-shailesh** [comment]
Implementation significantly improved by making use of datastores module.

**Comment by appukuttan-shailesh** [comment]
Nopes, that was a mistake (some test results got saved into the git directory!). Removed now.

**Pull request merged**